File: jlox/Lox.java
package jlox;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.charset.Charset;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.List;

/**
 * Lox
 * 
 * @apiNote Error code follow UNIX `sysexits.h`
 *          https://www.freebsd.org/cgi/man.cgi?query=sysexits&apropos=0&sektion=0&manpath=FreeBSD+4.3-RELEASE&format=html
 */
public class Lox {
	static boolean hadError = false;

	public static void main(String[] args) throws IOException {
		if (args.length > 1) {
			System.out.println("Usage: jlox [script]");
			System.exit(64);
		} else if (args.length == 1) {
			runFile(args[0]);
		} else {
			runPrompt();
		}
	}

	/**
	 * If you start jlox from the command line and give it a path to a file, it
	 * reads the file and executes it.
	 */
	private static void runFile(String path) throws IOException {
		byte[] bytes = Files.readAllBytes(Paths.get(path));
		run(new String(bytes, Charset.defaultCharset()));
		// We’ll use this to ensure we don’t try to execute code that has a known error.
		if (hadError) {
			System.exit(65);
		}
	}

	private static void runPrompt() throws IOException {
		InputStreamReader input = new InputStreamReader(System.in);
		BufferedReader reader = new BufferedReader(input);

		for (;;) {
			System.out.print("> ");
			String line = reader.readLine();
			// To kill an interactive command-line app, you usually type Control-D. Doing so
			// signals an “end-of-file” condition to the program.
			if (line == null) {
				break;
			}
			run(line);
			// If the user makes a mistake, it shouldn’t kill their entire session.
			hadError = false;
		}
	}

	private static void run(String source) {
		Scanner scanner = new Scanner(source);
		List<Token> tokens = scanner.scanTokens();

		// For now, just print the tokens.
		for (Token token : tokens) {
			System.out.println(token);
		}
	}

	static void error(int line, String message) {
		report(line, "", message);
	}

	private static void report(int line, String where, String message) {
		System.err.println("[line " + line + "] Error" + where + ": " + message);
		hadError = true;
	}
}

File: tslox/Lox.ts
import { readLines } from "https://deno.land/std@0.76.0/io/bufio.ts";
import { Scanner } from "./Scanner.ts";

export class Lox {
  private static hadError = false;

  private static run(source: string): void {
    const scanner = new Scanner(source);
    const tokens = scanner.scanTokens();

    // For now, just print the tokens.
    tokens.forEach((token) => console.log(token.toString()));
  }

  private static report(line: number, where: string, message: string): void {
    console.log(`[${line} line] Error ${where} : ${message}`);
    this.hadError = true;
  }

  static error(line: number, message: string): void {
    this.report(line, "", message);
  }

  /**
   * If you start tslox from the command line and give it a path to a file, it
   * reads the file and executes it.
   */
  private static async runFile(path: string) {
    const text = await Deno.readTextFile(path);
    this.run(text);

    if (this.hadError) {
      Deno.exit(65);
    }
  }

  private static async runPrompt() {
    for await (const line of readLines(Deno.stdin)) {
      this.run(line);
      this.hadError = false;
    }
  }

  public static async main() {
    if (Deno.args.length > 1) {
      console.log("Usage: tslox [script]");
      Deno.exit(64);
    } else if (Deno.args.length == 1) {
      await this.runFile(Deno.args[0]);
    } else {
      await this.runPrompt();
    }
  }
}

// Start the program
Lox.main()
  .then((a) => console.log(a))
  .catch((err) => console.error(err));
